xilem_core/src/deferred.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright 2024 the Xilem Authors
+// SPDX-License-Identifier: Apache-2.0
+
+use core::{fmt::Display, marker::PhantomData};
+
+use alloc::{boxed::Box, sync::Arc};
+
+use crate::{DynMessage, Message, NoElement, View, ViewId, ViewPathTracker};
+
+/// A `Context` for a [`View`](crate::View) implementation which supports
+/// asynchronous message reporting.
+pub trait AsyncCtx: ViewPathTracker {
+    /// Get a [`Proxy`] for this context.
+    // TODO: Maybe store the current path within this Proxy?
+    fn proxy(&mut self) -> Arc<dyn RawProxy>;
+}
+
+/// A handle to a Xilem driver which can be used to queue a message for a View.
+///
+/// These messages are [`crate::DynMessage`]s, which are sent to a view at
+/// a specific path.
+///
+/// This can be used for asynchronous event handling.
+/// For example, to get the result of a `Future` or a channel into
+/// the view, which then will ultimately.
+///
+/// In the Xilem crate, this will wrap an `EventLoopProxy` from Winit.
+///
+/// ## Lifetimes
+///
+/// It is valid for a [`Proxy`] to outlive the [`View`](crate::View) it is associated with.
+pub trait RawProxy: Send + Sync + 'static {
+    /// Send a `message` to the view at `path` in this driver.
+    ///
+    /// Note that it is only valid to send messages to views which expect
+    /// them, of the type they expect.
+    /// It is expected for [`View`](crate::View)s to panic otherwise, and the routing
+    /// will prefer to send stable.
+    ///
+    /// # Errors
+    ///
+    /// This method may error if the driver is no longer running, and in any other
+    /// cases directly documented on the context which was used to create this proxy.
+    /// It may also fail silently.
+    // TODO: Do we want/need a way to asynchronously report errors back to the caller?
```

Comment:
I guess that would be quite nice, so that apps, could show some indicator in the ui, when it failed.

Reply:
Yeah, there is a possible need to gracefully handle the stale messages.

At the moment, the future does get cancelled when the message would go stale, so there's only a very small window where this could error - practically, I think it's not really needed.

Reply:
Yeah we can see how long we can get away without this. Though it's probably not relevant for this PR